File: matrix/48rotateImage/rotateImage.java
import java.util.*;

public class rotateImage {
    public static int[][] rotate(int[][] matrix) {
        int matrixLength = matrix.length;
        for (int row = 0; row < matrixLength; row++) {
            for (int col = row + 1; col < matrixLength; col++) {
                int temp = matrix[row][col];
                matrix[row][col] = matrix[col][row];
                matrix[col][row] = temp;
            }
        }
        for (int row = 0; row < matrixLength; row++) {
            int l = 0,
                    r = matrixLength - 1;
            while (l < r) {
                int temp = matrix[row][l];
                matrix[row][l] = matrix[row][r];
                matrix[row][r] = temp;
                l++;
                r--;
            }
        }
        return matrix;
    }

    public static void main(String[] args) {
        int[][] matrix = { { 1, 2, 3, 4 }, { 5, 6, 7, 8 }, { 9, 10, 11, 12 }, { 13, 14, 15, 16 } };
        matrix = rotate(matrix);
        for (int i = 0; i < matrix.length; i++) {
            System.out.println(Arrays.toString(matrix[i]));
        }
    }
}

File: matrix/48rotateImage/rotateImage.py
class Solution:
    def rotate(self, matrix: list[list[int]]) -> list[list[int]]:
        matrixLength = len(matrix)
        for row in range(matrixLength):
            for col in range(row, matrixLength):
                matrix[row][col], matrix[col][row] = matrix[col][row], matrix[row][col]
        for row in range(len(matrix)):
            l, r = 0, matrixLength - 1
            while l < r:
                matrix[row][l], matrix[row][r] = matrix[row][r], matrix[row][l]
                l += 1
                r -= 1
        return matrix
print(Solution().rotate([[1,2,3,4], [5,6,7,8], [9,10,11,12], [13,14,15,16]]))